File: scripts/count_lines.py
"""Count lines of code in the project, grouped by extension.

Walks the repo, skipping node_modules, .next, .git, build artefacts, and lock
files. Counts total lines, blank lines, and comment-ish lines per extension
and prints a single summary table plus a grand total.
"""
from __future__ import annotations

import os
from collections import defaultdict

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))

SKIP_DIRS = {
    "node_modules",
    ".next",
    ".git",
    ".vercel",
    "dist",
    "build",
    "out",
    ".turbo",
    "coverage",
    "user_read_only_context",
}

EXTS = {
    ".ts",
    ".tsx",
    ".js",
    ".jsx",
    ".mjs",
    ".cjs",
    ".css",
    ".scss",
    ".sql",
    ".json",
    ".md",
    ".py",
}

def is_blank(line: str) -> bool:
    return line.strip() == ""

def is_comment(line: str, ext: str) -> bool:
    s = line.strip()
    if not s:
        return False
    if ext in {".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs", ".css", ".scss"}:
        return s.startswith("//") or s.startswith("/*") or s.startswith("*") or s.startswith("*/")
    if ext == ".sql":
        return s.startswith("--") or s.startswith("/*")
    if ext == ".py":
        return s.startswith("#")
    return False

per_ext_lines = defaultdict(int)
per_ext_blank = defaultdict(int)
per_ext_comment = defaultdict(int)
per_ext_files = defaultdict(int)

for dirpath, dirnames, filenames in os.walk(ROOT):
    # Skip unwanted directories in-place so os.walk doesn't descend
    dirnames[:] = [d for d in dirnames if d not in SKIP_DIRS and not d.startswith(".")]
    for name in filenames:
        ext = os.path.splitext(name)[1].lower()
        if ext not in EXTS:
            continue
        # Skip lockfiles
        if name in {"package-lock.json", "pnpm-lock.yaml", "bun.lockb", "yarn.lock"}:
            continue
        path = os.path.join(dirpath, name)
        try:
            with open(path, "r", encoding="utf-8", errors="ignore") as fh:
                lines = fh.readlines()
        except OSError:
            continue
        per_ext_files[ext] += 1
        per_ext_lines[ext] += len(lines)
        for line in lines:
            if is_blank(line):
                per_ext_blank[ext] += 1
            elif is_comment(line, ext):
                per_ext_comment[ext] += 1

def code(ext: str) -> int:
    return per_ext_lines[ext] - per_ext_blank[ext] - per_ext_comment[ext]

print(f"{'Ext':<7} {'Files':>7} {'Total':>10} {'Blank':>10} {'Comment':>10} {'Code':>10}")
print("-" * 58)
total_files = total_lines = total_blank = total_comment = total_code = 0
for ext in sorted(per_ext_lines.keys(), key=lambda e: -per_ext_lines[e]):
    files = per_ext_files[ext]
    lines = per_ext_lines[ext]
    blank = per_ext_blank[ext]
    comment = per_ext_comment[ext]
    c = code(ext)
    total_files += files
    total_lines += lines
    total_blank += blank
    total_comment += comment
    total_code += c
    print(f"{ext:<7} {files:>7} {lines:>10,} {blank:>10,} {comment:>10,} {c:>10,}")

print("-" * 58)
print(f"{'TOTAL':<7} {total_files:>7} {total_lines:>10,} {total_blank:>10,} {total_comment:>10,} {total_code:>10,}")

File: scripts/count-lines.mjs
// Count lines of code across the repo, broken down by extension.
// Skips node_modules, .next, .git, and other build artefacts.
import { promises as fs } from 'node:fs'
import path from 'node:path'

const ROOT = path.resolve(process.cwd())
const IGNORE_DIRS = new Set([
  'node_modules',
  '.next',
  '.git',
  '.vercel',
  'dist',
  'build',
  'coverage',
  '.turbo',
  'user_read_only_context',
])
const EXTS = new Set(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.css', '.sql', '.json', '.md'])

const stats = {}
let totalFiles = 0
let totalLines = 0

async function walk(dir) {
  const entries = await fs.readdir(dir, { withFileTypes: true })
  for (const entry of entries) {
    const full = path.join(dir, entry.name)
    if (entry.isDirectory()) {
      if (IGNORE_DIRS.has(entry.name) || entry.name.startsWith('.')) continue
      await walk(full)
      continue
    }
    const ext = path.extname(entry.name).toLowerCase()
    if (!EXTS.has(ext)) continue
    // Skip package-lock.json / pnpm-lock etc.
    if (entry.name.endsWith('lock.json') || entry.name === 'pnpm-lock.yaml') continue

    const content = await fs.readFile(full, 'utf8')
    const lines = content.split('\n').length
    if (!stats[ext]) stats[ext] = { files: 0, lines: 0 }
    stats[ext].files += 1
    stats[ext].lines += lines
    totalFiles += 1
    totalLines += lines
  }
}

await walk(ROOT)

// Print a clean summary
const rows = Object.entries(stats)
  .sort((a, b) => b[1].lines - a[1].lines)
  .map(([ext, s]) => `  ${ext.padEnd(6)} ${String(s.files).padStart(4)} files   ${String(s.lines).padStart(7)} lines`)

console.log('[v0] Dermaspace codebase summary')
console.log('---------------------------------')
console.log(rows.join('\n'))
console.log('---------------------------------')
console.log(`  TOTAL  ${String(totalFiles).padStart(4)} files   ${String(totalLines).padStart(7)} lines`)
